internals/secrethub/client_factory.go

```diff
@@ -36,31 +36,32 @@ func (f *clientFactory) Register(r FlagRegisterer) {
 
 // NewClient returns a new client that is configured to use the remote that
 // is set with the flag.
-func (f *clientFactory) NewClient() (secrethub.Client, error) {
+func (f *clientFactory) NewClient() (*secrethub.Client, error) {
 	if f.client == nil {
+		var credentialProvider credentials.Provider
 		if f.UseAWS {
-			client, err := secrethub.NewClientAWS(f.NewClientOptions())
-			if err != nil {
-				return nil, err
-			}
-			f.client = client
+			credentialProvider = credentials.UseAWS()
 		} else {
-			credential, err := f.store.Get()
-			if err != nil {
-				return nil, err
-			}
-			f.client = secrethub.NewClient(credential, auth.NewHTTPSigner(credential), f.NewClientOptions())
+			credentialProvider = f.store.Provider()
 		}
+
+		options := f.baseClientOptions()
+		options = append(options, secrethub.WithCredentials(credentialProvider))
+
+		client, err := secrethub.NewClient(options...)
+		if err != nil {
+			return nil, err
+		}
+		f.client = client
 	}
 	return f.client, nil
 }
 
-// NewClientOptions returns the client options configured by the flags.
-func (f *clientFactory) NewClientOptions() *secrethub.ClientOptions {
-	var opts secrethub.ClientOptions
+func (f *clientFactory) baseClientOptions() []secrethub.ClientOption {
```

Comment:
What differentiates these options as "base" options, compared to the option which credential provider to use, which is also configured on this factory?

Reply:
In later commits, a function is added that creates an unauthenticated client. This needs the base options, but not the option to use a credential.